contracts/RemovalQueue.sol

```diff
@@ -0,0 +1,177 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity =0.8.13;
+import {RemovalUtils} from "./RemovalUtils.sol";
+import "@openzeppelin/contracts-upgradeable/access/AccessControlEnumerableUpgradeable.sol";
+import "./Removal.sol";
+
+struct RemovalQueueByVintage {
+  mapping(uint16 => EnumerableSetUpgradeable.UintSet) queueByVintage;
+  uint16 earliestYear;
+  uint16 latestYear;
+}
+
+library RemovalQueue {
+  using RemovalUtils for uint256;
+  using EnumerableSetUpgradeable for EnumerableSetUpgradeable.UintSet;
+
+  uint16 private constant _DEFAULT_EARLIEST_YEAR = 2**16 - 1;
+  uint16 private constant _DEFAULT_LATEST_YEAR = 0;
```

Comment:
1 thoughts for caution: unless you know what you're doing here, leave uints at uint256 so that we can do a single pass later to optimize for gas around variable sizing/ordering later. IT's counterintuitive, but sometime using uint256 is cheaper

- https://dev.to/javier123454321/solidity-gas-optimizations-pt-3-packing-structs-23f4
- https://yos.io/2021/05/17/gas-efficient-solidity/

Reply:
> Using smaller-size uints such as uint8 is only more efficient when you can pack variables into the same storage slot, such as in structs. In other cases and in loops, uint256 is more gas efficient than uint8.

Since this is the case here (ealiestYear and latestYear are member of the RemovalQueueByVintage struct which is kept in storage), I think my implementation is the right call.
Really cool detail though, good callout.

Reply:
Also, I think it's funny that you misspelled `uint256` and uint16 in your comments (uin instead of uint), because I do that 100% of the time when writing these types. Something about that `t` just makes my brain skip over it.

Reply:
What about in the rest of the file where the same pattern is followed and we need to declare variables as non-uint256?

Reply:
Sorry, not sure I follow. Are you saying what if we need to instantiate variables for vintages outside of storage? Then I suppose we can just use uint256, the two types can be implicitly converted.

Reply:
This is more what I was looking for:



| test case | Using `uint16`  | Using `uint256` |
| ------------- | ------------- | ------------- |
| should mint a certificate with some of a single removal in round robin order and update state variables  | (801165 gas) |  (823110 gas) |
| should mint a certificate with one removal per supplier in round robin order and update state variables  | (1313367 gas) | (1292108 gas)  |